MeshLib/Properties.h

```diff
@@ -0,0 +1,220 @@
+/**
+ * \file
+ * \brief  Definition of the class Properties that implements a container of
+ *         properties.
+ *
+ * \copyright
+ * Copyright (c) 2012-2014, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#ifndef PROPERTIES_H_
+#define PROPERTIES_H_
+
+#include <cstdlib>
+#include <string>
+#include <map>
+
+#include <boost/any.hpp>
+#include <boost/optional.hpp>
+
+#include "logog/include/logog.hpp"
+
+#include "Location.h"
+
+#include "PropertyVector.h"
+
+namespace MeshLib
+{
+
+/// @brief Property manager on mesh items.
+/// Class Properties manages scalar, vector or matrix properties. For instance
+/// in groundwater flow porosity is a scalar property and permeabilty can be
+/// stored as a tensor property. Properties are assigned to mesh items, i.e.
+/// Node or Element objects. The createNewPropertyVector() method first creates a
+/// PropertyVector of template type T (scalar, vector or matrix).
+/// This class stores the PropertyVector, accessible by a combination of the
+/// name and the type of the mesh item (Node or Element).
+class Properties
+{
+public:
+	/// Method creates a PropertyVector if a PropertyVector with the same name
+	/// and the same type T was not already created before. In case there exists
+	/// already such a PropertyVector the returned boost::optional holds an
+	/// invalid/unusable PropertyVector.
+	/// There are two versions of this method. This method is used
+	/// when every mesh item at hand has its own property value, i.e. \f$n\f$
+	/// mesh item and \f$n\f$ different property values.
+	/// The user has to ensure the correct usage of the vector later on.
+	/// @tparam T type of the property value
+	/// @param name the name of the property
+	/// @param mesh_item_type for instance node or element assigned properties
+	/// @return On success a reference to a PropertyVector packed into a
+	///   boost::optional else an empty boost::optional.
+	template <typename T>
+	boost::optional<PropertyVector<T> &>
+	createNewPropertyVector(std::string const& name, MeshItemType mesh_item_type)
```

Comment:
I made class Properties a friend class of class PropertyVector and hid the constructors in protected scope of the class. So PropertyVector can only be constructed from Properties and not from the user.